File: Genral_All_Controls/PanAndZoomExample/PanAndZoomExample/ImageElement.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using Telerik.WinControls.UI;

namespace PanAndZoomExample
{
    public class ImageElement : LightVisualElement
    {
        private SizeF localOffset;
        private Point mouseDownLocation;
        private SizeF offset;
        private bool flag = false;
        private float zoom = 20f;

        protected override void InitializeFields()
        {
            base.InitializeFields();

            this.ImageLayout = System.Windows.Forms.ImageLayout.Stretch;
        }

        public SizeF Offset
        {
            get
            {
                return offset;
            }
            set
            {
                offset = value;
            }
        }

        protected override void OnMouseEnter(EventArgs e)
        {
            base.OnMouseEnter(e);

            this.ElementTree.Control.Cursor = Cursors.Hand;
        }

        protected override void OnMouseLeave(EventArgs e)
        {
            base.OnMouseLeave(e);

            this.ElementTree.Control.Cursor = Cursors.Default;
        }

        protected override void OnMouseDown(MouseEventArgs e)
        {
            base.OnMouseDown(e);

            this.localOffset = this.Offset;
            this.mouseDownLocation = e.Location;
            this.Capture = true;
            this.flag = true;
        }

        protected override void OnMouseMove(MouseEventArgs e)
        {
            base.OnMouseMove(e);

            if (e.Button == MouseButtons.Left && this.flag)
            {
                float width = this.localOffset.Width + e.Location.X - this.mouseDownLocation.X;
                float height = this.localOffset.Height + e.Location.Y - this.mouseDownLocation.Y;

                if (width > 0)
                {
                    width = 0;
                }

                if (height > 0)
                {
                    height = 0;
                }

                if (width < this.Parent.Size.Width - this.Size.Width)
                {
                    width = this.Parent.Size.Width - this.Size.Width;
                }

                if (height < this.Parent.Size.Height - this.Size.Height)
                {
                    height = this.Parent.Size.Height - this.Size.Height;
                }

                this.Offset = new SizeF(width, height);
                this.Parent.InvalidateArrange();
            }
        }

        protected override void OnMouseUp(MouseEventArgs e)
        {
            base.OnMouseUp(e);

            this.flag = false;
            this.Capture = false;
        }

        protected override void OnMouseWheel(MouseEventArgs e)
        {
            base.OnMouseWheel(e);

            if (Control.ModifierKeys == Keys.Control)
            {
                zoom += (e.Delta / 120);

                if (zoom < 1f)
                {
                    zoom = 1f;
                }

                if (zoom > 40f)
                {
                    zoom = 40f;
                }

                this.Parent.InvalidateMeasure(true);
                this.Parent.UpdateLayout();

                if (this.Size.Width > this.Parent.Size.Width)
                {
                    if (this.Offset.Width + this.Size.Width < this.Parent.Size.Width)
                    {
                        this.Offset = new SizeF(this.Offset.Width + (this.Parent.ControlBoundingRectangle.Right -
                             this.ControlBoundingRectangle.Right), this.Offset.Height);
                    }
                }

                if (this.Size.Height > this.Parent.Size.Height)
                {
                    if (this.Offset.Height + this.Size.Height < this.Parent.Size.Height)
                    {
                        this.Offset = new SizeF(this.Offset.Width, this.Offset.Height +
                             this.Parent.ControlBoundingRectangle.Bottom - this.ControlBoundingRectangle.Bottom);
                    }
                }

                this.Parent.InvalidateArrange(true);
            }
        }

        protected override SizeF MeasureOverride(SizeF availableSize)
        {
            if (Image != null)
                return new SizeF(this.Image.Width * zoom / 20f, this.Image.Height * zoom / 20f);

            return new SizeF(this.Parent.Size.Width, this.Parent.Size.Height);
        }
    }
}

File: Genral_All_Controls/PanAndZoomExample/PanAndZoomExampleVB/ImageElement.vb
Imports Telerik.WinControls.UI

Public Class ImageElement
    Inherits LightVisualElement

    Private localOffset As SizeF
    Private mouseDownLocation As Point
    Private _offset As SizeF
    Private flag As Boolean = False
    Private zoom As Single = 20.0F

    Protected Overrides Sub InitializeFields()
        MyBase.InitializeFields()
        Me.ImageLayout = System.Windows.Forms.ImageLayout.Stretch
    End Sub

    Public Property Offset As SizeF
        Get
            Return _offset
        End Get
        Set(ByVal value As SizeF)
            _offset = value
        End Set
    End Property

    Protected Overrides Sub OnMouseEnter(ByVal e As EventArgs)
        MyBase.OnMouseEnter(e)
        Me.ElementTree.Control.Cursor = Cursors.Hand
    End Sub

    Protected Overrides Sub OnMouseLeave(ByVal e As EventArgs)
        MyBase.OnMouseLeave(e)
        Me.ElementTree.Control.Cursor = Cursors.[Default]
    End Sub

    Protected Overrides Sub OnMouseDown(ByVal e As MouseEventArgs)
        MyBase.OnMouseDown(e)
        Me.localOffset = Me._offset
        Me.mouseDownLocation = e.Location
        Me.Capture = True
        Me.flag = True
    End Sub

    Protected Overrides Sub OnMouseMove(ByVal e As MouseEventArgs)
        MyBase.OnMouseMove(e)

        If e.Button = MouseButtons.Left AndAlso Me.flag Then
            Dim width As Single = Me.localOffset.Width + e.Location.X - Me.mouseDownLocation.X
            Dim height As Single = Me.localOffset.Height + e.Location.Y - Me.mouseDownLocation.Y

            If width > 0 Then
                width = 0
            End If

            If height > 0 Then
                height = 0
            End If

            If width < Me.Parent.Size.Width - Me.Size.Width Then
                width = Me.Parent.Size.Width - Me.Size.Width
            End If

            If height < Me.Parent.Size.Height - Me.Size.Height Then
                height = Me.Parent.Size.Height - Me.Size.Height
            End If

            Me._offset = New SizeF(width, height)
            Me.Parent.InvalidateArrange()
        End If
    End Sub

    Protected Overrides Sub OnMouseUp(ByVal e As MouseEventArgs)
        MyBase.OnMouseUp(e)
        Me.flag = False
        Me.Capture = False
    End Sub

    Protected Overrides Sub OnMouseWheel(ByVal e As MouseEventArgs)
        MyBase.OnMouseWheel(e)

        If Control.ModifierKeys = Keys.Control Then
            zoom += (e.Delta / 120)

            If zoom < 1.0F Then
                zoom = 1.0F
            End If

            If zoom > 40.0F Then
                zoom = 40.0F
            End If

            Me.Parent.InvalidateMeasure(True)
            Me.Parent.UpdateLayout()

            If Me.Size.Width > Me.Parent.Size.Width Then

                If Me._offset.Width + Me.Size.Width < Me.Parent.Size.Width Then
                    Me._offset = New SizeF(Me._offset.Width + (Me.Parent.ControlBoundingRectangle.Right - Me.ControlBoundingRectangle.Right), Me._offset.Height)
                End If
            End If

            If Me.Size.Height > Me.Parent.Size.Height Then

                If Me._offset.Height + Me.Size.Height < Me.Parent.Size.Height Then
                    Me._offset = New SizeF(Me._offset.Width, Me._offset.Height + Me.Parent.ControlBoundingRectangle.Bottom - Me.ControlBoundingRectangle.Bottom)
                End If
            End If

            Me.Parent.InvalidateArrange(True)
        End If
    End Sub

    Protected Overrides Function MeasureOverride(ByVal availableSize As SizeF) As SizeF
        If Image IsNot Nothing Then Return New SizeF(Me.Image.Width * zoom / 20.0F, Me.Image.Height * zoom / 20.0F)
        Return New SizeF(Me.Parent.Size.Width, Me.Parent.Size.Height)
    End Function
End Class